### PR TITLE
bug(opm): use db dir when modifying existing index

### PR DESCRIPTION
**Pull request by djzager**
**Description of the change:**

This makes it possible to add to an index image using `--from-index`.

**Motivation for the change:**

Squash :bug: 

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

Fixes: #221

### CODE REVIEW

File: pkg/lib/indexer/indexer.go [hunk 1]
```diff
@@ -79,6 +79,7 @@ func (i ImageIndexer) AddToIndex(request AddToIndexRequest) error {
 			}
 
 			databaseFile = path.Join(workingDir, dbLocation)
+			workingDir = path.Join(workingDir, path.Dir(dbLocation))
```
Comment (djzager) [inline pkg/lib/indexer/indexer.go:82]: I'm concerned about the `defer os.RemoveAll(workingDir)` but I wanted to get something up so I had time to think about test cases.

File: pkg/lib/indexer/indexer.go [hunk 2]
```diff
@@ -258,6 +221,33 @@ func (i ImageIndexer) buildDockerfile(binarySourceImage, workingDir, tag string)
 	return nil
 }
 
+func (i ImageIndexer) getDatabaseFile(workingDir, fromIndex string) (string, error) {
+	if fromIndex == "" {
+		return path.Join(workingDir, defaultDatabaseFile), nil
```
Comment (ecordell) [inline pkg/lib/indexer/indexer.go:226]: This is the actual fix, the rest is refactoring, correct?

**Comment by djzager** [comment]
/hold 

running into issues locally after rebase off master

**Comment by djzager** [comment]
I see now that `test/e2e/opm_test.go` is a good place to put tests for this. Is it alright @ecordell if I create an issue for this and add tests in a follow-up PR?

Before change:

```
+ /home/dzager/go/src/github.com/operator-framework/operator-registry/bin/opm index add -f docker.io/djzager/operator-index:latest -b docker.io/djzager/b-operator:v1.0.0,docker.io/djzager/b-operator:v1.0.1 --tag docker.io/djzager/operator-index:latest
INFO[0000] building the index                            bundles="[docker.io/djzager/b-operator:v1.0.0 docker.io/djzager/b-operator:v1.0.1]"
...
INFO[0018] [podman build -f index.Dockerfile305553182 -t docker.io/djzager/operator-index:latest .]  bundles="[docker.io/djzager/b-operator:v1.0.0 docker.io/djzager/b-operator:v1.0.1]"
ERRO[0036] STEP 1: FROM quay.io/operator-framework/upstream-registry-builder
STEP 2: LABEL operators.operatorframework.io.index.database.v1=/database/index.db
--> Using cache 149594f593de7aacf38c212c74fb33abab48f4b85b969d7d5d3f2433440aa559
STEP 3: ADD index_tmp751118410 /database
Error: error building at STEP "ADD index_tmp751118410 /database": error copying "/home/dzager/go/src/github.com/djzager/operator-bundles/index_tmp751118410" to "/home/dzager/.local/share/containers/storage/overlay/af257b74301a013fd7743613c0f7b0c6ff6fbb0dc9ccd4deb7c1d4e7beb90d2b/merged/database": Error processing tar file(exit status 1): open /build/.wh..wh..opq: invalid argument  bundles="[docker.io/djzager/b-operator:v1.0.0 docker.io/djzager/b-operator:v1.0.1]"
Error: error building image: STEP 1: FROM quay.io/operator-framework/upstream-registry-builder
STEP 2: LABEL operators.operatorframework.io.index.database.v1=/database/index.db
--> Using cache 149594f593de7aacf38c212c74fb33abab48f4b85b969d7d5d3f2433440aa559
STEP 3: ADD index_tmp751118410 /database
Error: error building at STEP "ADD index_tmp751118410 /database": error copying "/home/dzager/go/src/github.com/djzager/operator-bundles/index_tmp751118410" to "/home/dzager/.local/share/containers/storage/overlay/af257b74301a013fd7743613c0f7b0c6ff6fbb0dc9ccd4deb7c1d4e7beb90d2b/merged/database": Error processing tar file(exit status 1): open /build/.wh..wh..opq: invalid argument
. exit status 125
```

`opm index rm` before change:

```
$ /home/dzager/go/src/github.com/operator-framework/operator-registry/bin/opm index rm -f docker.io/djzager/operator-index:latest -o b -t docker.io/djzager/operator-index:latest
INFO[0000] building the index                            operators="[b]"
...
INFO[0018] [podman build -f index.Dockerfile968625881 -t docker.io/djzager/operator-index:latest .]  operators="[b]"
ERRO[0036] STEP 1: FROM quay.io/operator-framework/upstream-registry-builder
STEP 2: LABEL operators.operatorframework.io.index.database.v1=/database/index.db
--> Using cache 149594f593de7aacf38c212c74fb33abab48f4b85b969d7d5d3f2433440aa559
STEP 3: ADD index_tmp803276751 /database
Error: error building at STEP "ADD index_tmp803276751 /database": error copying "/home/dzager/go/src/github.com/djzager/operator-bundles/index_tmp803276751" to "/home/dzager/.local/share/containers/storage/overlay/7b70f1272a9376140525c43f2f4b182c81feed141260a6afd8604e64f1a6be36/merged/database": Error processing tar file(exit status 1): open /build/.wh..wh..opq: invalid argument  operators="[b]"
```

After change:

```
+ /home/dzager/go/src/github.com/operator-framework/operator-registry/bin/opm index add -f docker.io/djzager/operator-index:latest -b docker.io/djzager/b-operator:v1.0.0,docker.io/djzager/b-operator:v1.0.1 --tag docker.io/djzager/operator-index:latest
INFO[0000] building the index                            bundles="[docker.io/djzager/b-operator:v1.0.0 docker.io/djzager/b-operator:v1.0.1]"
...
INFO[0019] [podman build -f index.Dockerfile091376087 -t docker.io/djzager/operator-index:latest .]  bundles="[docker.io/djzager/b-operator:v1.0.0 docker.io/djzager/b-operator:v1.0.1]"
```

Verifying that the built index functions as expected:

```
$ podman run -d --name index -p 50051 docker.io/djzager/operator-index:latest
6b905e057e8a54626785df91d5c6ffef7067d7159604584281301a4c0dd83225

$ podman ps
CONTAINER ID  IMAGE                                    COMMAND               CREATED       STATUS           PORTS                     NAMES
6b905e057e8a  docker.io/djzager/operator-index:latest  registry serve --...  1 second ago  Up 1 second ago  0.0.0.0:37735->50051/tcp  index

$ grpcurl -plaintext localhost:37735 api.Registry/ListPackages
{
  "name": "a"
}
{
  "name": "b"
}
```

Verify that removing an operator from the index functions as expected:

```
$ /home/dzager/go/src/github.com/operator-framework/operator-registry/bin/opm index rm -f docker.io/djzager/operator-index:latest -o b -t docker.io/djzager/operator-index:latest
INFO[0000] building the index                            operators="[b]"
INFO[0000] Pulling previous image docker.io/djzager/operator-index:latest to get metadata  operators="[b]"
INFO[0000] running podman pull                           operators="[b]"
INFO[0002] Getting label data from previous image        operators="[b]"
INFO[0002] running podman inspect                        operators="[b]"
INFO[0002] running podman pull                           operators="[b]"
INFO[0004] running podman save                           operators="[b]"
INFO[0016] deleting packages                             pkg=b
INFO[0016] input has been sanitized                      pkg=b
INFO[0016] packages: [b]                                 pkg=b
INFO[0016] Generating dockerfile                         operators="[b]"
INFO[0016] writing dockerfile: index.Dockerfile650889418  operators="[b]"
INFO[0016] running podman build                          operators="[b]"
INFO[0016] [podman build -f index.Dockerfile650889418 -t docker.io/djzager/operator-index:latest .]  operators="[b]"

$ podman run -d --name index -p 50051 docker.io/djzager/operator-index:latest
4256b0bea2effcb9d6b87e0545b2b564e7cf206d051ffde8559ab27a8ad65994

$ grpcurl -plaintext localhost:40799 api.Registry/ListPackages
{
  "name": "a"
}
```

File: pkg/lib/indexer/indexer.go (hunk 2 above)
Comment (djzager) [reply pkg/lib/indexer/indexer.go:225]: The real fix is to make sure the working directory is adjacent to the database file (lines 68 and 124) `workingDir := path.Dir(databaseFile)`. The rest is refactoring yes.

Comment (djzager) [reply pkg/lib/indexer/indexer.go:225]: Only refactored because I need the same thing for `filter`.

**Comment by djzager** [comment]
Updated to include an e2e test: adding from an index.

**Pull request merged**